### PR TITLE
Async transfer API added

### PR DESCRIPTION
**Pull request by dmitry-zakablukov**
Good day.

This pr suggests some API updates to support async transfers.

Async transfer usage:
```
let mut transfer = rusb::Transfer::<'a>::new(&device_handle, 0)?;
transfer.set_transfer_type(rusb::constants::LIBUSB_TRANSFER_TYPE_CONTROL)?;
transfer.set_timeout(1000)?;
transfer.set_endpoint(0)?;

transfer.set_callback(Some(Box::new(move |status, data| {
    // process data
})))?;

transfer.submit_transfer(
    [0u8;10], // data
    0, bm_request_type,
    0, b_request,
    0, w_value,
    0, w_index,
)?;
```

You can take this changes as a base for your own API or just merge it as is.

### CODE REVIEW

File: src/transfer/unhandled_transfer.rs [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+
+        Ok(())
+    }
+
+    pub fn cancel_transfer(&mut self) -> Result<()> {
+        try_unsafe!(libusb_cancel_transfer(self.handle));
+        Ok(())
+    }
+
+    fn allocate_trhansfer_handle(iso_packets: i32) -> Result<*mut libusb_transfer> {
```
Comment (matallui) [inline src/transfer/unhandled_transfer.rs:106]: typo?

Comment (dmitry-zakablukov) [reply src/transfer/unhandled_transfer.rs]: Yes, thank you, fixed.

**Comment by a1ien** [comment]
I think it's good to split this PR.
Move log commit to separate PR.

**Comment by dmitry-zakablukov** [comment]
@a1ien , done: #76

**Comment by samueldotj** [comment]
@a1ien are you planning to merge this patch?

**Comment by a1ien** [comment]
I want find some time to rework. And maybe as separate crate in workspace. Because I want more test.
I promise that I find time until mid-autumn to check and rewrite and made some async support in rusb.

P.S. I am now in relocation process so to much new things in new country :)

**Comment by a1ien** [comment]
Closed in favor of #143

**Pull request closed without merging**